File: src/open.zig
const bun = @import("root").bun;
const string = bun.string;
const Output = bun.Output;
const Global = bun.Global;
const Environment = bun.Environment;
const strings = bun.strings;
const MutableString = bun.MutableString;
const stringZ = bun.stringZ;
const default_allocator = bun.default_allocator;
const C = bun.C;
const std = @import("std");
const DotEnv = @import("env_loader.zig");

const opener = switch (@import("builtin").target.os.tag) {
    .macos => "/usr/bin/open",
    .windows => "start",
    else => "xdg-open",
};

fn fallback(url: string) void {
    Output.prettyln("-> {s}", .{url});
    Output.flush();
}

pub fn openURL(url: stringZ) void {
    if (comptime Environment.isWasi) return fallback(url);

    var args_buf = [_]stringZ{ opener, url };

    maybe_fallback: {
        switch (bun.spawnSync(&.{
            .argv = &args_buf,

            .envp = null,

            .stderr = .inherit,
            .stdout = .inherit,
            .stdin = .inherit,

            .windows = if (Environment.isWindows) .{
                .loop = bun.JSC.EventLoopHandle.init(bun.JSC.MiniEventLoop.initGlobal(null)),
            } else {},
        }) catch break :maybe_fallback) {
            // don't fallback:
            .result => |*result| if (result.isOK()) return,
            .err => {},
        }
    }

    fallback(url);
}

pub const Editor = enum(u8) {
    none,
    sublime,
    vscode,
    atom,
    textmate,
    intellij,
    webstorm,
    vim,
    neovim,
    emacs,
    other,

    const StringMap = std.EnumMap(Editor, string);
    const StringArrayMap = std.EnumMap(Editor, []const [:0]const u8);

    const name_map = bun.ComptimeStringMap(Editor, .{
        .{ "sublime", .sublime },
        .{ "subl", .sublime },
        .{ "vscode", .vscode },
        .{ "code", .vscode },
        .{ "textmate", .textmate },
        .{ "mate", .textmate },
        .{ "atom", .atom },
        .{ "idea", .intellij },
        .{ "webstorm", .webstorm },
        .{ "nvim", .neovim },
        .{ "neovim", .neovim },
        .{ "vim", .vim },
        .{ "vi", .vim },
        .{ "emacs", .emacs },
    });

    pub fn byName(name: string) ?Editor {
        if (strings.indexOfChar(name, ' ')) |i| {
            return name_map.get(name[0..i]);
        }

        return name_map.get(name);
    }

    pub fn detect(env: *DotEnv.Loader) ?Editor {
        const vars = .{ "EDITOR", "VISUAL" };
        inline for (vars) |name| {
            if (env.get(name)) |value| {
                const basename = std.fs.path.basename(value);
                if (byName(basename)) |editor| {
                    return editor;
                }
            }
        }

        return null;
    }

    const which = @import("./which.zig").which;
    pub fn byPATH(env: *DotEnv.Loader, buf: *bun.PathBuffer, cwd: string, out: *[]const u8) ?Editor {
        const PATH = env.get("PATH") orelse return null;

        inline for (default_preference_list) |editor| {
            if (bin_name.get(editor)) |path| {
                if (which(buf, PATH, cwd, path)) |bin| {
                    out.* = bun.asByteSlice(bin);
                    return editor;
                }
            }
        }

        return null;
    }

    pub fn byPATHForEditor(env: *DotEnv.Loader, editor: Editor, buf: *bun.PathBuffer, cwd: string, out: *[]const u8) bool {
        const PATH = env.get("PATH") orelse return false;

        if (bin_name.get(editor)) |path| {
            if (path.len > 0) {
                if (which(buf, PATH, cwd, path)) |bin| {
                    out.* = bun.asByteSlice(bin);
                    return true;
                }
            }
        }

        return false;
    }

    pub fn byFallbackPathForEditor(editor: Editor, out: ?*[]const u8) bool {
        if (bin_path.get(editor)) |paths| {
            for (paths) |path| {
                if (std.fs.cwd().openFile(path, .{})) |opened| {
                    opened.close();
                    if (out != null) {
                        out.?.* = bun.asByteSlice(path);
                    }
                    return true;
                } else |_| {}
            }
        }

        return false;
    }

    pub fn byFallback(env: *DotEnv.Loader, buf: *bun.PathBuffer, cwd: string, out: *[]const u8) ?Editor {
        inline for (default_preference_list) |editor| {
            if (byPATHForEditor(env, editor, buf, cwd, out)) {
                return editor;
            }

            if (byFallbackPathForEditor(editor, out)) {
                return editor;
            }
        }

        return null;
    }

    pub const default_preference_list = [_]Editor{
        .vscode,
        .sublime,
        .atom,
        .neovim,

        .webstorm,
        .intellij,
        .textmate,
        .vim,
    };

    pub const bin_name: StringMap = brk: {
        var map = StringMap{};

        map.put(.sublime, "subl");
        map.put(.vscode, "code");
        map.put(.atom, "atom");
        map.put(.textmate, "mate");
        map.put(.intellij, "idea");
        map.put(.webstorm, "webstorm");
        map.put(.vim, "vim");
        map.put(.neovim, "nvim");
        map.put(.emacs, "emacs");
        map.put(.other, "");
        break :brk map;
    };

    pub const bin_path: StringArrayMap = brk: {
        var map = StringArrayMap{};

        if (Environment.isMac) {
            map.put(.vscode, &.{
                "/Applications/Visual Studio Code.app/Contents/Resources/app/bin/code",
                "/Applications/VSCodium.app/Contents/Resources/app/bin/code",
            });
            map.put(
                .atom,
                &.{
                    "/Applications/Atom.app/Contents/Resources/app/atom.sh",
                },
            );
            map.put(
                .sublime,
                &.{
                    "/Applications/Sublime Text 4.app/Contents/SharedSupport/bin/subl",
                    "/Applications/Sublime Text 3.app/Contents/SharedSupport/bin/subl",
                    "/Applications/Sublime Text 2.app/Contents/SharedSupport/bin/subl",
                    "/Applications/Sublime Text.app/Contents/SharedSupport/bin/subl",
                },
            );
        }

        break :brk map;
    };

    pub fn isJetBrains(editor: Editor) bool {
        return switch (editor) {
            .intellij, .webstorm => true,
            else => false,
        };
    }

    pub fn open(
        editor: Editor,
        binary: string,
        file: []const u8,
        line: ?string,
        column: ?string,
        _: std.mem.Allocator,
    ) !void {
        var spawned = try default_allocator.create(SpawnedEditorContext);
        spawned.* = .{};
        var file_path_buf_stream = std.io.fixedBufferStream(&spawned.file_path_buf);
        var file_path_buf_writer = file_path_buf_stream.writer();
        var args_buf = &spawned.buf;
        errdefer default_allocator.destroy(spawned);

        var i: usize = 0;

        if (editor == .vim or editor == .emacs or editor == .neovim) {
            args_buf[0] = opener;
            i += 1;

            args_buf[i] = binary;
            i += 1;

            if (Environment.isMac) {
                args_buf[i] = "--args";
                i += 1;
            }
        }

        args_buf[i] = binary;
        i += 1;

        if (editor == .vscode and line != null and line.?.len > 0) {
            args_buf[i] = "--goto";

            i += 1;
        }

        switch (editor) {
            .sublime, .atom, .vscode, .webstorm, .intellij => {
                try file_path_buf_writer.writeAll(file);
                if (line) |line_| {
                    if (line_.len > 0) {
                        try file_path_buf_writer.print(":{s}", .{line_});

                        if (!editor.isJetBrains()) {
                            if (column) |col| {
                                if (col.len > 0)
                                    try file_path_buf_writer.print(":{s}", .{col});
                            }
                        }
                    }
                }
                if (file_path_buf_stream.pos > 0) {
                    args_buf[i] = file_path_buf_stream.getWritten();
                    i += 1;
                }
            },
            .textmate => {
                try file_path_buf_writer.writeAll(file);
                const file_path = file_path_buf_stream.getWritten();

                if (line) |line_| {
                    if (line_.len > 0) {
                        args_buf[i] = "--line";
                        i += 1;

                        try file_path_buf_writer.print("{s}", .{line_});

                        if (column) |col| {
                            if (col.len > 0)
                                try file_path_buf_writer.print(":{s}", .{col});
                        }

                        const line_column = file_path_buf_stream.getWritten()[file_path.len..];
                        if (line_column.len > 0) {
                            args_buf[i] = line_column;
                            i += 1;
                        }
                    }
                }

                if (file_path_buf_stream.pos > 0) {
                    args_buf[i] = file_path;
                    i += 1;
                }
            },
            else => {
                if (file.len > 0) {
                    try file_path_buf_writer.writeAll(file);
                    const file_path = file_path_buf_stream.getWritten();
                    args_buf[i] = file_path;
                    i += 1;
                }
            },
        }

        spawned.child_process = std.process.Child.init(args_buf[0..i], default_allocator);
        var thread = try std.Thread.spawn(.{}, autoClose, .{spawned});
        thread.detach();
    }
    const SpawnedEditorContext = struct {
        file_path_buf: [1024 + bun.MAX_PATH_BYTES]u8 = undefined,
        buf: [10]string = undefined,
        child_process: std.process.Child = undefined,
    };

    fn autoClose(spawned: *SpawnedEditorContext) void {
        defer bun.default_allocator.destroy(spawned);

        Global.setThreadName("Open Editor");
        spawned.child_process.spawn() catch return;
        _ = spawned.child_process.wait() catch {};
    }
};

pub const EditorContext = struct {
    editor: ?Editor = null,
    name: string = "",
    path: string = "",
    const Fs = @import("./fs.zig");

    pub fn openInEditor(this: *EditorContext, editor_: Editor, blob: []const u8, id: string, tmpdir: std.fs.Dir, line: string, column: string) void {
        _openInEditor(this.path, editor_, blob, id, tmpdir, line, column) catch |err| {
            if (editor_ != .other) {
                Output.prettyErrorln("Error {s} opening in {s}", .{ @errorName(err), @tagName(editor_) });
            }

            this.editor = Editor.none;
        };
    }

    fn _openInEditor(path: string, editor_: Editor, blob: []const u8, id: string, tmpdir: std.fs.Dir, line: string, column: string) !void {
        var basename_buf: [512]u8 = undefined;
        var basename = std.fs.path.basename(id);
        if (strings.endsWith(basename, ".bun") and basename.len < 499) {
            bun.copy(u8, &basename_buf, basename);
            basename_buf[basename.len..][0..3].* = ".js".*;
            basename = basename_buf[0 .. basename.len + 3];
        }

        try tmpdir.writeFile(basename, blob);

        var opened = try tmpdir.openFile(basename, .{});
        defer opened.close();
        var path_buf: bun.PathBuffer = undefined;
        try editor_.open(
            path,
            try bun.getFdPath(opened.handle, &path_buf),
            line,
            column,
            default_allocator,
        );
    }

    pub fn autoDetectEditor(this: *EditorContext, env: *DotEnv.Loader) void {
        if (this.editor == null) {
            this.detectEditor(env);
        }
    }
    pub fn detectEditor(this: *EditorContext, env: *DotEnv.Loader) void {
        var buf: bun.PathBuffer = undefined;

        var out: string = "";
        // first: choose from user preference
        if (this.name.len > 0) {
            // /usr/bin/vim
            if (std.fs.path.isAbsolute(this.name)) {
                this.editor = Editor.byName(std.fs.path.basename(this.name)) orelse Editor.other;
                this.path = this.name;
                return;
            }

            // "vscode"
            if (Editor.byName(std.fs.path.basename(this.name))) |editor_| {
                if (Editor.byPATHForEditor(env, editor_, &buf, Fs.FileSystem.instance.top_level_dir, &out)) {
                    this.editor = editor_;
                    this.path = Fs.FileSystem.instance.dirname_store.append(string, out) catch unreachable;
                    return;
                }

                // not in path, try common ones
                if (Editor.byFallbackPathForEditor(editor_, &out)) {
                    this.editor = editor_;
                    this.path = Fs.FileSystem.instance.dirname_store.append(string, out) catch unreachable;
                    return;
                }
            }
        }

        // EDITOR=code
        if (Editor.detect(env)) |editor_| {
            if (Editor.byPATHForEditor(env, editor_, &buf, Fs.FileSystem.instance.top_level_dir, &out)) {
                this.editor = editor_;
                this.path = Fs.FileSystem.instance.dirname_store.append(string, out) catch unreachable;
                return;
            }

            // not in path, try common ones
            if (Editor.byFallbackPathForEditor(editor_, &out)) {
                this.editor = editor_;
                this.path = Fs.FileSystem.instance.dirname_store.append(string, out) catch unreachable;
                return;
            }
        }

        // Don't know, so we will just guess based on what exists
        if (Editor.byFallback(env, &buf, Fs.FileSystem.instance.top_level_dir, &out)) |editor_| {
            this.editor = editor_;
            this.path = Fs.FileSystem.instance.dirname_store.append(string, out) catch unreachable;
            return;
        }

        this.editor = Editor.none;
    }
};

File: src/runtime/cli/open.rs
use std::io::Write as _;

use bun_core::Global;
use bun_core::{ZStr, strings};
use bun_dotenv as dot_env;
use bun_paths::{self, PathBuffer};
use bun_resolver::fs as Fs;
use bun_which::which;

use crate::api::bun::process::sync;

// ──────────────────────────────────────────────────────────────────────────

#[cfg(target_os = "macos")]
const OPENER: &[u8] = b"/usr/bin/open";
#[cfg(windows)]
const OPENER: &[u8] = b"start";
#[cfg(not(any(target_os = "macos", windows)))]
const OPENER: &[u8] = b"xdg-open";

// ──────────────────────────────────────────────────────────────────────────

#[repr(u8)]
#[derive(Copy, Clone, PartialEq, Eq, Hash, strum::IntoStaticStr, enum_map::Enum)]
#[strum(serialize_all = "snake_case")] // Vscode → "vscode"
pub enum Editor {
    None,
    Sublime,
    Vscode,
    Atom,
    Textmate,
    Intellij,
    Webstorm,
    Vim,
    Neovim,
    Emacs,
    Other,
}

// Note: `bin_name` is an `enum_map::EnumMap<E, Option<V>>` (sparse map);
// `bin_path` is a match-fn because of `#[cfg]` gating.

bun_core::comptime_string_map! {
    static NAME_MAP: Editor = {
        b"sublime" => Editor::Sublime,
        b"subl" => Editor::Sublime,
        b"vscode" => Editor::Vscode,
        b"code" => Editor::Vscode,
        b"textmate" => Editor::Textmate,
        b"mate" => Editor::Textmate,
        b"atom" => Editor::Atom,
        b"idea" => Editor::Intellij,
        b"webstorm" => Editor::Webstorm,
        b"nvim" => Editor::Neovim,
        b"neovim" => Editor::Neovim,
        b"vim" => Editor::Vim,
        b"vi" => Editor::Vim,
        b"emacs" => Editor::Emacs,
    };
}

impl Editor {
    pub(crate) fn by_name(name: &[u8]) -> Option<Editor> {
        if let Some(i) = strings::index_of_char(name, b' ') {
            return NAME_MAP.get(&name[0..i as usize]).copied();
        }
        NAME_MAP.get(name).copied()
    }

    pub(crate) fn detect(env: &mut dot_env::Loader) -> Option<Editor> {
        const VARS: [&[u8]; 2] = [b"EDITOR", b"VISUAL"];
        for name in VARS {
            if let Some(value) = env.get(name) {
                let basename = bun_paths::basename(value);
                if let Some(editor) = Self::by_name(basename) {
                    return Some(editor);
                }
            }
        }
        None
    }

    /// On a hit, the binary path is `buf[..len]`.
    pub(crate) fn by_path_for_editor(
        env: &mut dot_env::Loader,
        editor: Editor,
        buf: &mut PathBuffer,
        cwd: &[u8],
    ) -> Option<usize> {
        let path_env = env.get(b"PATH")?;

        if let Some(bin_name) = BIN_NAME[editor] {
            if !bin_name.is_empty() {
                if let Some(bin) = which(buf, path_env, cwd, bin_name) {
                    return Some(bin.len());
                }
            }
        }

        None
    }

    pub(crate) fn by_fallback_path_for_editor(
        editor: Editor,
        out: Option<&mut &'static [u8]>,
    ) -> bool {
        if let Some(paths) = bin_path(editor) {
            for path in paths {
                match bun_sys::File::open_at(bun_sys::Fd::cwd(), path, bun_sys::O::RDONLY, 0) {
                    bun_sys::Result::Ok(opened) => {
                        let _ = opened.close(); // close error is non-actionable
                        if let Some(out) = out {
                            *out = path.as_bytes();
                        }
                        return true;
                    }
                    bun_sys::Result::Err(_) => {}
                }
            }
        }

        false
    }

    pub(crate) fn by_fallback<'a>(
        env: &mut dot_env::Loader,
        buf: &'a mut PathBuffer,
        cwd: &[u8],
    ) -> Option<(Editor, &'a [u8])> {
        for &editor in &DEFAULT_PREFERENCE_LIST {
            if let Some(len) = Self::by_path_for_editor(env, editor, buf, cwd) {
                return Some((editor, &buf[..len]));
            }

            let mut static_out: &'static [u8] = b"";
            if Self::by_fallback_path_for_editor(editor, Some(&mut static_out)) {
                return Some((editor, static_out));
            }
        }

        None
    }

    pub(crate) fn is_jet_brains(self) -> bool {
        matches!(self, Editor::Intellij | Editor::Webstorm)
    }

    pub(crate) fn open(
        self,
        binary: &[u8],
        file: &[u8],
        line: Option<&[u8]>,
        column: Option<&[u8]>,
    ) -> crate::Result<()> {
        let mut argv: Vec<Box<[u8]>> = Vec::with_capacity(10);
        macro_rules! push_arg {
            ($s:expr) => {{
                argv.push(Box::<[u8]>::from(&$s[..]));
            }};
        }

        if matches!(self, Editor::Vim | Editor::Emacs | Editor::Neovim) {
            push_arg!(OPENER);
            push_arg!(binary);

            #[cfg(target_os = "macos")]
            {
                push_arg!(b"--args");
            }
        }

        push_arg!(binary);

        if self == Editor::Vscode && line.is_some() && !line.unwrap().is_empty() {
            push_arg!(b"--goto");
        }

        match self {
            Editor::Sublime
            | Editor::Atom
            | Editor::Vscode
            | Editor::Webstorm
            | Editor::Intellij => {
                let mut file_path: Vec<u8> = Vec::with_capacity(file.len() + 16);
                file_path.extend_from_slice(file);
                if let Some(line_) = line {
                    if !line_.is_empty() {
                        write!(file_path, ":{}", bstr::BStr::new(line_))
                            .map_err(|_| crate::Error::WriteFailed)?;

                        if !self.is_jet_brains() {
                            if let Some(col) = column {
                                if !col.is_empty() {
                                    write!(file_path, ":{}", bstr::BStr::new(col))
                                        .map_err(|_| crate::Error::WriteFailed)?;
                                }
                            }
                        }
                    }
                }
                if !file_path.is_empty() {
                    argv.push(file_path.into_boxed_slice());
                }
            }
            Editor::Textmate => {
                let mut line_column: Vec<u8> = Vec::new();
                if let Some(line_) = line {
                    if !line_.is_empty() {
                        push_arg!(b"--line");

                        write!(line_column, "{}", bstr::BStr::new(line_))
                            .map_err(|_| crate::Error::WriteFailed)?;

                        if let Some(col) = column {
                            if !col.is_empty() {
                                write!(line_column, ":{}", bstr::BStr::new(col))
                                    .map_err(|_| crate::Error::WriteFailed)?;
                            }
                        }
                    }
                }

                let has_line_column = !line_column.is_empty();
                if has_line_column {
                    argv.push(line_column.into_boxed_slice());
                }

                if !file.is_empty() || has_line_column {
                    push_arg!(file);
                }
            }
            _ => {
                if !file.is_empty() {
                    push_arg!(file);
                }
            }
        }

        // bun_threading has no detached-spawn helper; std::thread::spawn is used
        // and the JoinHandle is dropped, detaching the thread.
        std::thread::Builder::new()
            .spawn(move || auto_close(argv))
            .map_err(|_| crate::Error::ThreadSpawnFailed)?;
        Ok(())
    }
}

const DEFAULT_PREFERENCE_LIST: [Editor; 8] = [
    Editor::Vscode,
    Editor::Sublime,
    Editor::Atom,
    Editor::Neovim,
    Editor::Webstorm,
    Editor::Intellij,
    Editor::Textmate,
    Editor::Vim,
];

static BIN_NAME: std::sync::LazyLock<enum_map::EnumMap<Editor, Option<&'static [u8]>>> =
    std::sync::LazyLock::new(|| {
        enum_map::EnumMap::from_fn(|k| match k {
            Editor::Sublime => Some(&b"subl"[..]),
            Editor::Vscode => Some(&b"code"[..]),
            Editor::Atom => Some(&b"atom"[..]),
            Editor::Textmate => Some(&b"mate"[..]),
            Editor::Intellij => Some(&b"idea"[..]),
            Editor::Webstorm => Some(&b"webstorm"[..]),
            Editor::Vim => Some(&b"vim"[..]),
            Editor::Neovim => Some(&b"nvim"[..]),
            Editor::Emacs => Some(&b"emacs"[..]),
            Editor::Other => Some(&b""[..]),
            Editor::None => None,
        })
    });

fn bin_path(editor: Editor) -> Option<&'static [&'static ZStr]> {
    #[cfg(target_os = "macos")]
    {
        // `const { &[...] }` forces const-promotion so the array lives in
        // `'static` storage (otherwise `&[..]` borrows a stack temporary).
        match editor {
            Editor::Vscode => Some(
                const {
                    &[
                ZStr::from_static(b"/Applications/Visual Studio Code.app/Contents/Resources/app/bin/code\0"),
                ZStr::from_static(b"/Applications/VSCodium.app/Contents/Resources/app/bin/code\0"),
            ]
                },
            ),
            Editor::Atom => Some(
                const {
                    &[ZStr::from_static(
                        b"/Applications/Atom.app/Contents/Resources/app/atom.sh\0",
                    )]
                },
            ),
            Editor::Sublime => {
                Some(
                    const {
                        &[
                ZStr::from_static(b"/Applications/Sublime Text 4.app/Contents/SharedSupport/bin/subl\0"),
                ZStr::from_static(b"/Applications/Sublime Text 3.app/Contents/SharedSupport/bin/subl\0"),
                ZStr::from_static(b"/Applications/Sublime Text 2.app/Contents/SharedSupport/bin/subl\0"),
                ZStr::from_static(b"/Applications/Sublime Text.app/Contents/SharedSupport/bin/subl\0"),
            ]
                    },
                )
            }
            _ => None,
        }
    }
    #[cfg(not(target_os = "macos"))]
    {
        let _ = editor;
        None
    }
}

fn auto_close(argv: Vec<Box<[u8]>>) {
    Global::set_thread_name(bun_core::zstr!("Open Editor"));

    // FIXME(windows-leak): the sync::spawn path
    // requires a `WindowsOptions.loop_`; `MiniEventLoop::init_global` heap-allocates a
    // MiniEventLoop + uv_loop_t into a thread-local that is NEVER torn down. Because this
    // runs on a fresh detached std::thread per `Editor::open()` call, every editor-open on
    // Windows leaks one MiniEventLoop + uv_loop_t (+ DotEnv Loader/Map if env was null).
    // Proper fix needs either (a) a MiniEventLoop teardown helper (none exists today), or
    // (b) plumbing the caller's existing EventLoopHandle through `Editor::open`
    // (signature change to Editor::open + callers). Both are out-of-scope for this file.
    let _ = sync::spawn(&sync::Options {
        argv,
        envp: None,
        stderr: sync::SyncStdio::Inherit,
        stdout: sync::SyncStdio::Inherit,
        stdin: sync::SyncStdio::Inherit,
        #[cfg(windows)]
        windows: crate::api::bun::process::WindowsOptions {
            loop_: bun_jsc::EventLoopHandle::init_mini(bun_event_loop::MiniEventLoop::init_global(
                None, None,
            )),
            ..Default::default()
        },
        ..Default::default()
    });
}

// ──────────────────────────────────────────────────────────────────────────

pub struct EditorContext {
    pub(crate) editor: Option<Editor>,
    // Note: `name`/`path` are never freed; `path` is backed by
    // `Fs.FileSystem.instance.dirname_store` (process-lifetime arena) or aliases `name`.
    pub name: &'static [u8],
    pub path: &'static [u8],
}

impl Default for EditorContext {
    fn default() -> Self {
        Self {
            editor: None,
            name: b"",
            path: b"",
        }
    }
}

impl EditorContext {
    /// `detect_editor` records `Editor::None` when nothing was found so the
    /// search is not repeated; to callers that means "no editor".
    pub(crate) fn found(&self) -> Option<Editor> {
        self.editor.filter(|e| *e != Editor::None)
    }

    pub(crate) fn auto_detect_editor(&mut self, env: &mut dot_env::Loader) {
        if self.editor.is_none() {
            self.detect_editor(env);
        }
    }

    pub(crate) fn detect_editor(&mut self, env: &mut dot_env::Loader) {
        let mut buf = PathBuffer::uninit();
        let top_level_dir = Fs::FileSystem::get().top_level_dir;
        let dirname_store = Fs::FileSystem::get().dirname_store;

        // first: choose from user preference
        if !self.name.is_empty() {
            // /usr/bin/vim
            if bun_paths::is_absolute(self.name) {
                self.editor =
                    Some(Editor::by_name(bun_paths::basename(self.name)).unwrap_or(Editor::Other));
                self.path = self.name;
                return;
            }

            // "vscode"
            if let Some(editor_) = Editor::by_name(bun_paths::basename(self.name)) {
                if let Some(len) = Editor::by_path_for_editor(env, editor_, &mut buf, top_level_dir)
                {
                    self.editor = Some(editor_);
                    self.path = dirname_store
                        .append_slice(&buf[..len])
                        .expect("unreachable");
                    return;
                }

                // not in path, try common ones
                let mut static_out: &'static [u8] = b"";
                if Editor::by_fallback_path_for_editor(editor_, Some(&mut static_out)) {
                    self.editor = Some(editor_);
                    self.path = dirname_store.append_slice(static_out).expect("unreachable");
                    return;
                }
            }
        }

        // EDITOR=code
        if let Some(editor_) = Editor::detect(env) {
            if let Some(len) = Editor::by_path_for_editor(env, editor_, &mut buf, top_level_dir) {
                self.editor = Some(editor_);
                self.path = dirname_store
                    .append_slice(&buf[..len])
                    .expect("unreachable");
                return;
            }

            // not in path, try common ones
            let mut static_out: &'static [u8] = b"";
            if Editor::by_fallback_path_for_editor(editor_, Some(&mut static_out)) {
                self.editor = Some(editor_);
                self.path = dirname_store.append_slice(static_out).expect("unreachable");
                return;
            }
        }

        // Don't know, so we will just guess based on what exists
        if let Some((editor_, out)) = Editor::by_fallback(env, &mut buf, top_level_dir) {
            self.editor = Some(editor_);
            self.path = dirname_store.append_slice(out).expect("unreachable");
            return;
        }

        self.editor = Some(Editor::None);
    }
}
